api/services/bulk-docs.js

```diff
@@ -26,45 +19,96 @@ const checkForDuplicates = docs => {
   }
 };
 
-const generateBatches = (docs, batchSize) => {
+const generateBatches = (ids, batchSize) => {
   const batches = [];
-  while (docs.length > 0) {
-    const batch = docs.splice(0, batchSize);
+  while (ids.length > 0) {
+    const batch = ids.splice(0, batchSize);
     batches.push(batch);
   }
   return batches;
 };
 
-const generateBatchPromise = (batch, res, { isFinal } = {}) => {
-  return db.medic.bulkDocs(batch).then(result => {
-    let resString = JSON.stringify(result);
-    resString += isFinal ? '' : ',';
-    res.write(resString);
+const fetchDocs = ids => {
+  if (ids.length === 0) {
+    return Promise.resolve([]);
+  }
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(extractDocs);
+};
+
+const incrementAttemptCounts = (docs, attemptCounts) => {
+  docs.forEach(doc => {
+    attemptCounts[doc._id] = (attemptCounts[doc._id] || 0) + 1;
   });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
-      })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+const deleteDocs = (docsToDelete, deletionAttemptCounts, updateAttemptCounts, cumulativeResult = []) => {
+  const finishedModifications = cumulativeResult.map(docUpdate => docUpdate.id);
+  const parentMap = {};
+  let docsToUpdate;
+  let docsToModify;
+  docsToDelete.forEach(doc => doc._deleted = true);
+  incrementAttemptCounts(docsToDelete, deletionAttemptCounts);
+  return utils.updateParentContacts(docsToDelete, parentMap)
+    .then(updatedParents => {
+      docsToUpdate = updatedParents;
```

Comment:
should we filter `updatedParents` though `docsToDelete`, because you might be trying to delete one of the parents? (in which case the delete is the more important operation)

Reply:
Yep good point! I meant to implement something like that but must have gotten sidetracked